cpp/source/deal.II/mp_values.templates.h

```diff
@@ -6,82 +6,273 @@
  */
 
 #include <cap/mp_values.h>
-#include <stdexcept>
+#include <cap/utils.h>
+#include <random>
 #include <tuple>
+#include <stdexcept>
 
 namespace cap
 {
 
+//////////////////////// COMPOSITE MAT /////////////////////////////////////////
 template <int dim, int spacedim>
-MPValues<dim, spacedim>::MPValues(
-    MPValuesParameters<dim, spacedim> const &params)
-    : materials()
+void CompositeMat<dim, spacedim>::get_values(std::string const &key,
+                                             active_cell_iterator const &cell,
+                                             std::vector<double> &values) const
+
+{
+  dealii::types::material_id const material_id = cell->material_id();
+  auto got = _materials.find(material_id);
+  if (got == _materials.end())
+    throw std::runtime_error("Invalid material id " +
+                             std::to_string(material_id));
+  auto material = got->second;
+  material->get_values(key, cell, values);
+}
+//////////////////////// COMPOSITE PRO /////////////////////////////////////////
+template <int dim, int spacedim>
+void CompositePro<dim, spacedim>::get_values(std::string const &key,
+                                             active_cell_iterator const &cell,
+                                             std::vector<double> &values) const
+{
+  auto got = _properties.find(key);
+  if (got == _properties.end())
+    throw std::runtime_error("Invalid material property " + key);
+  auto property = got->second;
+  property->get_values(key, cell, values);
+}
+//////////////////////// UNIFORM CONSTANT //////////////////////////////////////
+template <int dim, int spacedim>
+UniformConstantMPValues<dim, spacedim>::UniformConstantMPValues(
+    double const &val)
+    : _val(val)
+{
+}
+template <int dim, int spacedim>
+void UniformConstantMPValues<dim, spacedim>::get_values(
+    std::string const &key, active_cell_iterator const &cell,
+    std::vector<double> &values) const
+{
+  std::ignore = key;
+  std::ignore = cell;
+  std::fill(values.begin(), values.end(), _val);
+}
+//////////////////////// RANDOM DISTRIBUTION ///////////////////////////////////
+std::function<double(std::default_random_engine &)>
+_build_parameter(boost::property_tree::ptree const &parameter_database)
```

Comment:
I am not a big fan of having functions starting with _ It will make it harder to know if it's a private variable or a function.